pkg/mallochook/mallochook.c

```diff
@@ -0,0 +1,208 @@
+// References:
+// https://refspecs.linuxfoundation.org/elf/elf.pdf
+// - Section "Shared Object Dependencies" on the order of run-time symbol resolution
+
+#define _GNU_SOURCE 1
+
+#include <dlfcn.h>
+#include <malloc.h>
+#include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#if __GNUC__ > 4
+#  include <stdatomic.h>
+#  define mallochook_cnt_t atomic_size_t
+#  define mallochook_inc(a, v) a += v
+#  define mallochook_dec(a, v) a -= v
+#  define mallochook_get(a) a
+#else
+#  define mallochook_cnt_t unsigned long
+#  define mallochook_inc(a, v) __sync_fetch_and_add(&a, v)
+#  define mallochook_dec(a, v) __sync_fetch_and_sub(&a, v)
+#  define mallochook_get(a) __sync_fetch_and_add(&a, 0)
+#endif
+
+static void *(*mallochook_malloc)(size_t size);
+static void *(*mallochook_calloc)(size_t nmemb, size_t size);
+static void *(*mallochook_realloc)(void *ptr, size_t size);
+static void *(*mallochook_reallocarray)(void *ptr, size_t nmemb, size_t size);
+static void (*mallochook_free)(void *ptr);
+
+static int (*mallochook_posix_memalign)(void **memptr, size_t alignment, size_t size);
+static void *(*mallochook_aligned_alloc)(size_t alignment, size_t size);
+static void *(*mallochook_valloc)(size_t size);
+
+static void *(*mallochook_memalign)(size_t alignment, size_t size);
+static void *(*mallochook_pvalloc)(size_t size);
+
+static mallochook_cnt_t mallochook_heap_inuse;
+static mallochook_cnt_t mallochook_heap_alloc;
+
+static void mallochook_track_alloc(void *ptr) {
+    if (ptr != NULL) {
+        size_t usable = malloc_usable_size(ptr);
+        mallochook_inc(mallochook_heap_inuse, usable);
+        mallochook_inc(mallochook_heap_alloc, usable);
+    }
+}
+
+static void mallochook_track_free(void *ptr) {
+    if (ptr != NULL) {
+        size_t usable = malloc_usable_size(ptr);
+        mallochook_dec(mallochook_heap_inuse, usable);
+    }
+}
+
+static void* mallochook_loadsym(const char *name) {
+    dlerror(); // Clear last error, see dlsym man page
+    void *ptr = dlsym(RTLD_NEXT, name);
+    if (ptr == NULL) {
+        char *err = dlerror();
+        if (err == NULL) {
+            err = "symbol is defined, but null";
+        }
+        // printf calls malloc, but we may have just failed to load one
+        const char *reloc_err = "error patching symbol ";
+        // not much we can do about errors or partial writes here, but this
+        // keeps the -Werror happy.
+        int n = 0;
+        n += write(2, reloc_err, strlen(reloc_err));
+        n += write(2, name, strlen(name));
+        n += write(2, ": ", 2);
+        n += write(2, err, strlen(err));
+        n += write(2, "\n", 1);
+    }
+    return ptr;
+}
+
+// Temporary calloc implementation to use while we are loading symbols using
+// dlsym, which in turn calls calloc. This codepath can handle allocation errors
+// and will use a global static buffer instead. mallochook_init will ensure that
+// all symbols are resolved during process startup when only one thread is
```

Comment:
this refers to the fact that `mallochook_init` is called as a `constructor` below I assume?